storm-client/src/jvm/org/apache/storm/windowing/TimeEvictionPolicy.java

```diff
@@ -41,8 +47,10 @@ public TimeEvictionPolicy(int windowLength) {
     public Action evict(Event<T> event) {      
         long now = evictionContext == null ? System.currentTimeMillis() : evictionContext.getReferenceTime();
         long diff = now - event.getTimestamp();
-        if (diff >= windowLength) {
+        if (diff >= (windowLength + delta)) {
             return Action.EXPIRE;
+        } else if (diff < 0) { // do not process events beyond current ts
+            return Action.KEEP;
```

Comment:
yes, ideally events should be ordered since its processing time, unless `System.currentTimeMillis` goes backwards, so I assumed it could be out of order.